x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java

```diff
@@ -968,6 +968,19 @@ private Long getAnomalyMemoryRequirement(PersistentTask<?> task) {
         long currentlyNecessaryTier = nodeLoads.stream().mapToLong(NodeLoad::getAssignedJobMemory).sum();
         // The required NATIVE node memory is the largest job and our static overhead.
         long currentlyNecessaryNode = largestJob == 0 ? 0 : largestJob + MachineLearning.NATIVE_EXECUTABLE_CODE_OVERHEAD.getBytes();
+        // If we are using `auto` && have at least one job, that means our native node size should be at least native capacity provided
+        // via our `MINIMUM_AUTOMATIC_NODE_SIZE`. Otherwise, if we have to auto-calculate the JVM size, it could be much smaller than
+        // what will truly be used.
+        if (currentlyNecessaryNode > 0 && useAuto) {
+            currentlyNecessaryNode = Math.max(
+                currentlyNecessaryNode,
+                NativeMemoryCalculator.allowedBytesForMl(
+                    NativeMemoryCalculator.MINIMUM_AUTOMATIC_NODE_SIZE,
+                    maxMachineMemoryPercent,
+                    useAuto
+                )
+            );
+        }
```

Comment:
Putting this here seems very much like a sticking plaster.

It seems to me that we have a problem that's the reverse of what's documented in: https://github.com/elastic/elasticsearch/blob/a61088063e54942d39d83109e2db29fe70a27163/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/NativeMemoryCalculator.java#L180-L182

It seems like what we really want is a method in `NativeMemoryCalculator` that reliably calculates "given an amount of memory needed for ML processes, how big does the node need to be?"

Currently the methods in `NativeMemoryCalculator` seem to do this in a very convoluted way.  I have a feeling all this code could be simplified by taking a step back from all the low level methods we have at the moment and replacing them with a new set of methods whose arguments and return values more closely match the high level requirements we have.

For example, take this comment:

> Otherwise, if we have to auto-calculate the JVM size, it could be much smaller than what will truly be used.

To me that says, "there's a bug in the method that auto-calculates the JVM size".  I think it's because we're trying to calculate the JVM size independently of the node size when the two are linked.  And it's extremely hard because the formula for going from node size to JVM size is currently not reversible, so we have to breach encapsulation and think about what we'd want on Cloud for the ranges of input values where the inverse formula is not well defined.

I guess we can merge this now as the sticking plaster approach is at least low risk, but hopefully within the next few months we can make the formula for JVM size given node size reversible and then do a major refactor.

Reply:
@droberts195 exactly, this is "sticky plaster". But I only wanted to change scale down (thus reduce impact), and want to address this particular issue. 

I 100% agree, this needs to be refactored and reworked (especially in the calculator).